### ISSUE TITLE
[HOLD for payment 2023-03-14] [$1000] The text "Account ending in" is not translated despite having a translation in en.js and es.js

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15333]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Add a bank account with plaid on staging
2. Change the language to Spanish
3. Open the payments page

## Expected Result
The text "account ending in" should be in Spanish

## Actual Result
The text "account ending in" is in English

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.75-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
<img width="387" alt="Screenshot 2023-02-21 at 5 00 00 PM copy" src="https://user-images.githubusercontent.com/43996225/220452977-30bd2005-72af-4277-a290-1afd89d8c215.png">
<img width="960" alt="account ending in spanish" src="https://user-images.githubusercontent.com/43996225/220452994-8a8dd8ad-d16c-435e-bc0b-804d62e1a014.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @esh-g 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676980043749489

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019ab80a4989edacaf</li>
        <li>Upwork Job ID: 1628408075501514752</li>
        <li>Last Price Increase: 2023-02-22</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Current assignee @NicMendonca is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @flodnv (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Incorrect translation is being shown for payment method list.

### What is the root cause of that problem?
We're showing the payment method description as it is received from the backend. The backend however, is sending that in english. Thus, setting local language has no effect on what is displayed. This is a sample response that is received after adding a bank account:
```
[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"bankAccountList","value":{"1579091":{"title":"Plaid Checking","description":"Account ending in 0000","methodID":1579091,"key":"bankAccount-1579091","accountType":"bankAccount","accountData":{"accountNumber":"111122XXXXXX0000","additionalData":{"acceptTerms":true,"bankName":"","country":"US","currency":"USD","fieldsType":"local","plaidAccessToken":"","plaidAccountID":"MxMQgonJv6TEwzxA5jP3F6rLpBAgWpC7PQx4m","previousState":2},"addressName":"Plaid Checking","allowDebit":false,"bankAccountID":1579091,"created":"2022-06-03 20:26:43","defaultCredit":true,"expensifyCardDomain":"memorydump.games","isExpensifyCardSettlementAccount":true,"isSavings":false,"processor":"bancorp","riskChecked":false,"routingNumber":"011401533","sharees":["jeremy@memorydump.games","admin+2000@tulacostarica.com","marco@expensifail.com","allroundexperts+1@gmail.com"],"state":"OPEN","type":"PERSONAL"}}}}]
```

### What changes do you think we should make in order to solve the problem?

We can solve it either at the backend (by reading the users language set and sending translated version to the frontend).
If that is something we need to avoid, then we can run a `replace` function in the description column of the response to replace the `account ending in` key automatically.

### What alternative solutions did you explore? (Optional)
None

**Comment by MelvinBot** [comment]
📣 @allroundexperts! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
📣 @Milos213! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by allroundexperts** [comment]
Contributor details
Your Expensify account email: allroundexperts@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01ee96f841f75f9dd1

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by MelvinBot** [comment]
📣 @Prince-Mendiratta! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by allroundexperts** [comment]
## Updated Proposal

### Please re-state the problem that we are trying to solve in this issue.
Incorrect translation is being shown for payment method list.

### What is the root cause of that problem?
We're showing the payment method description as it is received from the backend. The backend however, is sending that in english. Thus, setting local language has no effect on what is displayed. This is a sample response that is received after adding a bank account:
```
[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"bankAccountList","value":{"1579091":{"title":"Plaid Checking","description":"Account ending in 0000","methodID":1579091,"key":"bankAccount-1579091","accountType":"bankAccount","accountData":{"accountNumber":"111122XXXXXX0000","additionalData":{"acceptTerms":true,"bankName":"","country":"US","currency":"USD","fieldsType":"local","plaidAccessToken":"","plaidAccountID":"MxMQgonJv6TEwzxA5jP3F6rLpBAgWpC7PQx4m","previousState":2},"addressName":"Plaid Checking","allowDebit":false,"bankAccountID":1579091,"created":"2022-06-03 20:26:43","defaultCredit":true,"expensifyCardDomain":"memorydump.games","isExpensifyCardSettlementAccount":true,"isSavings":false,"processor":"bancorp","riskChecked":false,"routingNumber":"011401533","sharees":["jeremy@memorydump.games","admin+2000@tulacostarica.com","marco@expensifail.com","allroundexperts+1@gmail.com"],"state":"OPEN","type":"PERSONAL"}}}}]
```

### What changes do you think we should make in order to solve the problem?

We can solve it either at the backend (by reading the users language set and sending translated version to the frontend).
If that is something we need to avoid, then we can run a `replace` function in the description column of the response to replace the `account ending in` key automatically. 

We can also just not use the description that comes with the backend. We can override it on the frontend. To make this useable everywhere, we can pass a new prop to to the function [here](https://github.com/allroundexperts/Expensify/blob/e109f78266887aa2ee9b77a83a1c129711a204a9/src/libs/PaymentUtils.js#L34) called `translate`. This will look like:

```
function formatPaymentMethods(bankAccountList, cardList, payPalMeData = null, translate = null) {
```
In the same function, we'll also need to add the following [here](https://github.com/allroundexperts/Expensify/blob/e109f78266887aa2ee9b77a83a1c129711a204a9/src/libs/PaymentUtils.js#L46).

```
...translate && {
                description: translate(`Account ending in ${bankAccount.accountData.accountNumber.slice(-4)}`),
            },
```
We can then pass on the `translate` argument to `formatPaymentMethods` everywhere its being used so that this works across the board.

We can do the same for credit card and paypal account description as well.

### What alternative solutions did you explore? (Optional)
None

**Comment by MelvinBot** [comment]
📣 @ggalletty! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by allroundexperts** [comment]
> @Prince-Mendiratta @allroundexperts thanks for your proposals! @Milos213 was first one so going with that

Thanks for the feedback @rushatgabhane!

**Comment by ggalletty** [comment]
> @ggalletty i love your solution as well because it's handling both LTR and RTL. But I don't think this is the right time to do it. We should probably do it after the entire app's audit is done for supporting RTL.
> 
> I would have approved your solution but @Milos213 was the first one, thanks so much for your proposal!

Thank you @rushatgabhane for feedback! Hopefully I will be able to contribute when the time comes 🙌

**Comment by flodnv** [comment]
Hmmm, good catch! @Milos213 your proposal makes sense. However, after checking, I think in this API call, the backend does not return a field containing only the last four / the account number, or does it...? If it does not, we should add a field 'lastFourDigits' and use that instead. Let me know what you think 🙇

**Comment by Milos213** [comment]
![image](https://user-images.githubusercontent.com/115748623/221577762-9a60ad92-555f-49b2-8911-823dc88c8805.png)

Can't we use this one which was already used in delete modal?
<img width="417" alt="delete" src="https://user-images.githubusercontent.com/115748623/221579271-07e47c2d-18ba-4876-b436-9316a5a71691.png">

**Comment by flodnv** [comment]
Yes! Does this work for payments _cards_ though?

**Comment by Milos213** [comment]
I think so. The field name is `cardNumber`. I was not able to test debit card but this code was already used in delete modal:
https://github.com/Expensify/App/blob/9d04b46456253955c96981db9ef38245953baf18/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L216

**Comment by flodnv** [comment]
Hmmm looking again at the API I think it will work, but please make sure to test this in your PR. I created a followup issue for myself to remove these untranslated fields from the API in a little while (after making sure they're unused): https://github.com/Expensify/Expensify/issues/264765

Go ahead and proceed with the PR @Milos213 ! 🙇

**Comment by MelvinBot** [comment]
📣 @Milos213 You have been assigned to this job by @flodnv!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.79-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15555


If no regressions arise, payment will be issued on 2023-03-14. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane / @flodnv] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane / @flodnv] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane / @flodnv] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@NicMendonca] Determine if we should create a regression test for the bug.
- [ ] [@rushatgabhane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to the [appropriate location](https://github.com/Expensify/App/blob/main/contributingGuides/TestRail_Navigation_C%2B.md) to ensure the same bug will not reach production again.
- [ ] [@NicMendonca] Review the proposed regression test steps and location.
- [ ] [@NicMendonca] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by NicMendonca** [comment]
@esh-g @rushatgabhane @Milos213 Can you please apply for the job so that we can issue payment? https://www.upwork.com/jobs/~019ab80a4989edacaf

Thanks!

**Comment by esh-g** [comment]
Applied under the name Tanya Gupta

**Comment by NicMendonca** [comment]
everyone has been paid! thank you!
